File: construct-binary-tree-from-inorder-and-postorder-traversal.py
__author__ = 'stephenosullivan'

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """

        def _buildTree(inorder, postorder):
            if postorder:
                if len(postorder) == 1:
                    return TreeNode(postorder[0])

                # check equality (long left branch)
                if inorder == postorder:
                    node = TreeNode(inorder[-1])
                    head = node
                    for i in inorder[-2::-1]:
                        node.left = TreeNode(i)
                        node = node.left
                    return head

                # check reverse (long right branch)
                if inorder == reversed(postorder):
                    node = TreeNode(postorder[-1])
                    head = node
                    for i in postorder[-2::-1]:
                        node.right = TreeNode(i)
                        node = node.right
                    return head

                # Partition into left and right branches
                node = TreeNode(postorder[-1])
                indexright = len(postorder) - 1
                # Find partition point
                while postorder[-1] != inorder[indexright]:
                    indexright -= 1
                node.right = _buildTree(inorder[indexright + 1:], postorder[indexright:-1])
                node.left = _buildTree(inorder[:indexright], postorder[:indexright])
                return node

        return _buildTree(inorder, postorder)

File: construct-binary-tree-from-inorder-and-postorder-traversal.cpp
//
//  construct-binary-tree-form-inorder-and-postorder-traversal.cpp
//
//
//  Created by Stephen O'Sullivan on 8/23/15.
//
//

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if (!postorder.empty()){
            vector<int>::iterator iter;
            vector<int>::reverse_iterator rev_iter;

            // Save computations by checking for series of left or right branches
            bool reverseCheck  = true;
            for(iter=inorder.begin(), rev_iter=postorder.rbegin(); iter!=inorder.end(), rev_iter!=postorder.rend(); ++iter, ++rev_iter){
                if(*iter!=*rev_iter){
                    reverseCheck=false;
                    break;
                }
            }

            if (inorder==postorder){
                TreeNode* node = new TreeNode(postorder.back());
                TreeNode* head = node;
                for(rev_iter=inorder.rbegin()+1;rev_iter!=inorder.rend();++rev_iter){
                    node->left = new TreeNode(*rev_iter);
                    node = node->left;
                }
                return head;
            }


            else if (reverseCheck){
                TreeNode* node = new TreeNode(postorder.back());
                TreeNode* head = node;
                for(rev_iter=postorder.rbegin()+1;rev_iter!=postorder.rend();++rev_iter){
                    node->right = new TreeNode(*rev_iter);
                    node = node->right;
                }
                return head;
            }

            // Need to partition
            TreeNode* node = new TreeNode(postorder.back());
            if (inorder.size() > 1){
                int index = find(inorder.begin(), inorder.end(), postorder.back()) - inorder.begin();
                postorder.pop_back();
                vector<int> inorderRightSub(inorder.begin()+index+1, inorder.end());
                vector<int> postorderRightSub(postorder.begin()+index, postorder.end());
                vector<int> inorderLeftSub(inorder.begin(), inorder.begin()+index);
                vector<int> postorderLeftSub(postorder.begin(), postorder.begin()+index);

                node->right = buildTree(inorderRightSub, postorderRightSub);
                node->left = buildTree(inorderLeftSub, postorderLeftSub);
            }
            return node;
        }
        else{
            return NULL;
        }

    }
};



void printNode(TreeNode* node){
    while (node != NULL) {
        printNode(node->left);
        cout << node->val << endl;;
        printNode(node->right);
        break;

    }
}

int main(){
    vector<int> inorder = {1,3,2};
    vector<int> postorder = {3,2,1};

    Solution sol;
    TreeNode* node = sol.buildTree(inorder, postorder);
    cout << node->val << endl;
    cout << node->right->val << endl;
    cout << node->right->left->val << endl;

    printNode(node);


}